pkg/networking/cilium/upgrade_checker.go

```diff
@@ -0,0 +1,118 @@
+package cilium
+
+import (
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/aws/eks-anywhere/pkg/cluster"
+)
+
+// InstallationUpgradeInfo contains information about a Cilium installation upgrade
+type InstallationUpgradeInfo struct {
+	DaemonSet ComponentUpgradeInfo
+	Operator  ComponentUpgradeInfo
+}
+
+// Needed determines if an upgrade is needed or not
+// Returns true if any of the installation components needs an upgrade
+func (c InstallationUpgradeInfo) Needed() bool {
+	return c.DaemonSet.Needed() || c.Operator.Needed()
+}
+
+// Reason returns the reason why an upgrade might be needed
+// If no upgrade needed, returns empty string
+// For multiple components with needed upgrades, it composes their reasons into one
+func (c InstallationUpgradeInfo) Reason() string {
+	if !c.Needed() {
+		return ""
+	}
+
+	s := make([]string, 0, 2)
+	if c.DaemonSet.Reason != "" {
+		s = append(s, c.DaemonSet.Reason)
+	}
+	if c.Operator.Reason != "" {
+		s = append(s, c.Operator.Reason)
+	}
+
+	return strings.Join(s, " - ")
+}
+
+// ComponentUpgradeInfo contains upgrade information for a Cilium component
+type ComponentUpgradeInfo struct {
+	Reason   string
+	OldImage string
+	NewImage string
+}
+
+// Needed determines if an upgrade is needed or not
+func (c ComponentUpgradeInfo) Needed() bool {
+	return c.Reason != ""
```

Comment:
Why use Reason than another field that specifies if it's latest or if it needs an upgrade or something?

Reply:
Not sure  I get the question. Are you saying why we don't have a specific field to indicate if upgrade is needed? If that's the case: I started with a boolean for that reason, but then `Reason` and that flag always were set at the same time so the flag seemed redundant.

If we ever want to add one, we just need to change this method, that's why I added (vs checking `c.Reason != ""` all over the code).

Reply:
Yea I was more wondering whether having a Reason always means an upgrade is needed, which in that case, would `UpgradeReason` make more sense? Maybe I am having a trouble time visualizing but I assume Reason is more general.

Reply:
Yes, `Reason` is only set when an upgrade is needed. I like `UpgradeReason`, changing